File: typescript/src/sopsEncrypt.ts
// encryptForSecret hand-implements the narrow slice of the SOPS encrypted-file
// format signetd requires: a single `value: <ciphertext>` leaf, encrypted to
// exactly one age (X25519) recipient. This exists because signetd never
// encrypts secrets on a caller's behalf — SyncBundle/TriggerSync both require
// content that is already real SOPS ciphertext (signet's documented trust
// model is "only SOPS ciphertext leaves the operator's machine") — and no npm
// package implements the actual SOPS envelope format (everything under
// "sops" on npm is either decrypt-only or produces an incompatible custom
// format). See bytepunx/signet-clients#49.
//
// The algorithm below was verified against the github.com/getsops/sops/v3
// source and cross-checked by round-tripping output through the real `sops`
// binary (see sopsEncrypt.test.ts) — follow it precisely; do not improvise.
import { createCipheriv, createHash, randomBytes } from "node:crypto";
import { Encrypter, armor } from "age-encryption";
import { Document, Scalar } from "yaml";

const SOPS_VERSION = "3.13.1";

/**
 * encryptForSecret produces a SOPS-compatible encrypted YAML document for a
 * single secret value, encrypted to the given age recipient (a `age1...`
 * public key string, e.g. from signetd's GetSOPSPublicKey RPC). The result is
 * byte-for-byte suitable as the `content` of a signet SyncBundle/TriggerSync
 * call — decrypting it with `sops --decrypt` (given the matching age
 * identity) recovers a document with exactly one field, `value`, holding the
 * original secret.
 *
 * value may be a UTF-8 string or raw bytes; either way it is encrypted and
 * MAC'd as-is (this mirrors what `sops --encrypt` produces for a file
 * containing exactly `value: "<secret>"`, always rendered as SOPS's `str`
 * leaf type).
 */
export async function encryptForSecret(publicKey: string, value: string | Uint8Array): Promise<Uint8Array> {
  const recipient = publicKey.trim();
  if (recipient === "") {
    throw new Error("signet: encryptForSecret: publicKey must not be empty");
  }

  const plaintext = typeof value === "string" ? Buffer.from(value, "utf8") : Buffer.from(value);

  const dataKey = randomBytes(32);

  const mac = createHash("sha512").update(plaintext).digest("hex").toUpperCase();

  const encryptedValue = gcmSealSopsEnc(dataKey, plaintext, "value:");

  // Whole-second precision, no fractional digits: sops parses lastmodified
  // back into a Go time.Time and re-derives the MAC's AAD by reformatting it
  // with Go's time.RFC3339 layout, which always truncates to whole seconds —
  // date.toISOString() always includes millisecond digits (".000Z"), so
  // leaving them in here would make the AAD sops recomputes at decrypt time
  // disagree with the one used below, and MAC decryption would fail
  // authentication.
  const lastmodified = new Date().toISOString().replace(/\.\d+Z$/, "Z");
  const encryptedMac = gcmSealSopsEnc(dataKey, Buffer.from(mac, "utf8"), lastmodified);

  const encrypter = new Encrypter();
  try {
    encrypter.addRecipient(recipient);
  } catch (err) {
    throw new Error(`signet: encryptForSecret: invalid age recipient: ${err instanceof Error ? err.message : String(err)}`);
  }
  const wrappedDataKey = await encrypter.encrypt(dataKey);
  const armoredDataKey = armor.encode(wrappedDataKey);

  const doc = new Document({
    value: encryptedValue,
    sops: {
      age: [
        {
          recipient,
          enc: armoredDataKey,
        },
      ],
      lastmodified,
      mac: encryptedMac,
      version: SOPS_VERSION,
    },
  });

  // lastmodified must round-trip as a plain string. Go's yaml.v3 (which sops
  // is built on) auto-detects an *unquoted* RFC3339 scalar as a time.Time
  // when decoding, which then fails to unmarshal into sops's Metadata.
  // LastModified string field ("expected type 'string', got unconvertible
  // type 'time.Time'") — force double-quoted style so it round-trips as a
  // string regardless of the parser's schema.
  const lastmodifiedNode = doc.getIn(["sops", "lastmodified"], true);
  if (lastmodifiedNode instanceof Scalar) {
    lastmodifiedNode.type = Scalar.QUOTE_DOUBLE;
  }

  return new TextEncoder().encode(doc.toString());
}

/**
 * gcmSealSopsEnc encrypts plaintext with AES-256-GCM under a fresh random
 * 32-byte nonce (not the standard 12 bytes — this matches sops's own
 * `cipher.NewGCMWithNonceSize(cipher, 32)`), authenticating aad, and renders
 * the result in sops's `ENC[...]` leaf syntax.
 */
function gcmSealSopsEnc(dataKey: Buffer, plaintext: Buffer, aad: string): string {
  const nonce = randomBytes(32);
  const cipher = createCipheriv("aes-256-gcm", dataKey, nonce);
  cipher.setAAD(Buffer.from(aad, "utf8"));
  const ciphertext = Buffer.concat([cipher.update(plaintext), cipher.final()]);
  const tag = cipher.getAuthTag();

  return `ENC[AES256_GCM,data:${ciphertext.toString("base64")},iv:${nonce.toString("base64")},tag:${tag.toString("base64")},type:str]`;
}

File: go/sops_encrypt.go
package signet

import (
	"crypto/rand"
	"fmt"

	sops "github.com/getsops/sops/v3"
	sopsaes "github.com/getsops/sops/v3/aes"
	sopsage "github.com/getsops/sops/v3/age"
	sopscommon "github.com/getsops/sops/v3/cmd/sops/common"
	sopsconfig "github.com/getsops/sops/v3/config"
	sopsyaml "github.com/getsops/sops/v3/stores/yaml"
	"gopkg.in/yaml.v3"
)

// sopsDataKeySize is the size, in bytes, of the random AES-256 data key SOPS
// generates per file — matches sops's own Tree.GenerateDataKeyWithKeyServices.
const sopsDataKeySize = 32

// sopsFileVersion is recorded in the file's "sops.version" metadata field.
// The real sops CLI only requires this to be non-empty to decrypt (it
// doesn't gate decryption on matching its own build version); pinned to the
// getsops/sops/v3 release this package builds against rather than importing
// that module's own version package, which pulls in a full CLI framework
// (github.com/urfave/cli) for a single string constant.
const sopsFileVersion = "3.13.1"

// EncryptForSecret produces SOPS-compatible encrypted YAML content for value,
// encrypted to publicKey (an age recipient string, e.g. as returned by
// GitOpsService's GetSOPSPublicKey RPC). The result is exactly the shape
// signetd's gitops sync path (TriggerSync, SyncBundle) and the `signet
// secret set` CLI both expect: a YAML document with a top-level "value" key,
// SOPS-encrypted to a single age recipient.
//
// This exists because SyncBundle stores secret file content as-is — it never
// encrypts on the server's behalf (see design/draft.md's "only SOPS
// ciphertext leaves the operator's machine" trust model in bytepunx/signet).
// A plaintext value submitted via SyncBundle fails to decrypt and is
// silently never stored; EncryptForSecret is the correct way to produce a
// value SyncBundle will actually persist. See bytepunx/signet-clients#49.
//
// No external sops binary is required: this uses the same
// github.com/getsops/sops/v3 + filippo.io/age libraries signetd's own
// server-side decrypt path already depends on, run in the encrypt
// direction — a library needing a binary on PATH is a worse story than a
// CLI needing one, which is why `signet secret set` (CLI-only) can get away
// with shelling out but this can't.
func EncryptForSecret(publicKey string, value []byte) ([]byte, error) {
	if publicKey == "" {
		return nil, fmt.Errorf("signet: public key must not be empty")
	}

	mk, err := sopsage.MasterKeyFromRecipient(publicKey)
	if err != nil {
		return nil, fmt.Errorf("signet: parse age recipient: %w", err)
	}

	dataKey := make([]byte, sopsDataKeySize)
	if _, err := rand.Read(dataKey); err != nil {
		return nil, fmt.Errorf("signet: generate data key: %w", err)
	}
	if err := mk.Encrypt(dataKey); err != nil {
		return nil, fmt.Errorf("signet: encrypt data key: %w", err)
	}

	plain, err := yaml.Marshal(map[string]string{"value": string(value)})
	if err != nil {
		return nil, fmt.Errorf("signet: marshal plaintext yaml: %w", err)
	}

	store := sopsyaml.NewStore(&sopsconfig.YAMLStoreConfig{})
	branches, err := store.LoadPlainFile(plain)
	if err != nil {
		return nil, fmt.Errorf("signet: load plaintext yaml: %w", err)
	}

	tree := sops.Tree{
		Branches: branches,
		Metadata: sops.Metadata{
			KeyGroups: []sops.KeyGroup{{mk}},
			Version:   sopsFileVersion,
		},
	}

	if err := sopscommon.EncryptTree(sopscommon.EncryptTreeOpts{
		DataKey: dataKey,
		Tree:    &tree,
		Cipher:  sopsaes.NewCipher(),
	}); err != nil {
		return nil, fmt.Errorf("signet: encrypt tree: %w", err)
	}

	encrypted, err := store.EmitEncryptedFile(tree)
	if err != nil {
		return nil, fmt.Errorf("signet: emit encrypted file: %w", err)
	}
	return encrypted, nil
}
